### ISSUE TITLE
Traversal fails inspite of increased heap size

### ISSUE THREAD

**Issue by Arun-KumarH** [arangodb/arangodb#7346]
## My Environment

* __ArangoDB Version__:       3.3.14
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start in Docker
* __Infrastructure__:               own
* __Operating System__:       Suse Leap
* __Used Package__:             Docker - official Docker library

## Component, Query & Data

__Affected feature__:
 Graphs traversal query

__AQL query (if applicable)__:
```info: request to method traversal over transport grpc error {"request":{"request":{"vertex":"start_vertex","start_vertex":"service_teams/9a2e582997494bee9066bbcf2aa52218","start_vertices":null,"opts":{"expander":"var connections = [];\n    config.datasource.getOutEdges(vertex).forEach(function (e) {\n      if(  (e._id.indexOf(\"loc_has_parent_loc\") > -1)  || (e._id.indexOf(\"loc_is_associated_with_org\") > -1)  || (e._id.indexOf(\"st_is_allocated_loc\") > -1) ) {\n        connections.push({ vertex: require(\"internal\").db._document(e._to), edge: e});\n      }\n    });\n    return connections;"},"collection_name":"","edge_name":"","data":true,"path":false}},"err":{"code":500,"message":"Invalid string length"}}```

__Problem__:
I have a traversal api which traverses through 3 vertices and returns around 300 document vertices.
For 200 I get the proper response but when I increase the number of vertices to 300, the traversal api throws an error `invalid string length`. I have increased the heap space for application to 8GB but as can be seen below the heap_used is way far too less. I am quite not sure if this issue is during serialization to JSON as there is sufficient heap memory available.

**Comment by Simran-B** [comment]
Can you share the actual traversal code? Maybe it's possible to replace it by AQL traversal?

**Comment by Arun-KumarH** [comment]
I am using the the graphs traversal api from arangojs:

import { Graph, Database } from "arangojs";
traversedData = await this.graph.traversal(vertex, opts);

Vertex is: service_teams/9a2e582997494bee9066bbcf2aa52218 
opts is: `{"expander":"var connections = [];\n    config.datasource.getOutEdges(vertex).forEach(function (e) {\n      if(  (e._id.indexOf(\"loc_has_parent_loc\") > -1)  || (e._id.indexOf(\"loc_is_associated_with_org\") > -1)  || (e._id.indexOf(\"st_is_allocated_loc\") > -1) ) {\n        connections.push({ vertex: require(\"internal\").db._document(e._to), edge: e});\n      }\n    });\n    return connections;"}`

Basically vertices are `service_teams`, `locations` and `organizations`, and edges are `st_is_allocated_loc`, `loc_has_parent_loc`, `loc_is_associated_with_org` : 
`(service_teams) ->st_is_allocated_loc -> (locations) ->loc_is_associated_with_org-> (organization)`
There is only one service_team and one organization and if I have 200 locations then the traversal works fine, if I increase the number of locations to 300 then the traversal fails with `Invalid string length` error.

**Comment by Arun-KumarH** [comment]
Also one more thing I noticed is although I had the expander opts to traverse only through 3 edges from ArangoDB log files i see that it actually traverses through entire graph and when returning the data I see the vertices only for those 3 nodes.  So is it better to return only the vertex_ids and then make individual queries? please suggest.

**Comment by Simran-B** [comment]
@Arun-KumarH I meant to ask for the output of the traversal, not the code (you posted the code in the first post already) - my bad!

Nonetheless, if I understand correctly, then you want to hop from a service_team vertex over edges in edge collection `st_is_allocated_loc` to location vertices and from there via edges in edge collection `loc_is_associated_with_org` to organization vertices. This should translate quite well to an AQL query:

```js
//import { Database, aql } from "arangojs";
const { Database, aql } = require("arangojs");
const db = new Database();

(async function() {
  const service_team = "9a2e582997494bee9066bbcf2aa52218"
  try {
    const cursor = await db.query(aql`
    FOR team IN service_teams
      FILTER team._key == ${service_team}
      FOR location, locationEdge IN OUTBOUND team st_is_allocated_loc
      FOR organization, organizationEdge IN OUTBOUND location loc_is_associated_with_org
        RETURN { team, location, locationEdge, organization, organizationEdge }
    `);
    let result = await cursor.all();
    console.log(`Count: ${result.length}`)
  } catch(err) {
    console.error(err);
  }
})();
```

If you don't need the full vertex/edge documents, then you can adjust the RETURN statement accordingly.

**Comment by Arun-KumarH** [comment]
Above query works fine, but with `_api/traversal` HTTP endpoint, I see that it traverses through entire graph although the edges were limited to 3.
Probably that was the reason for string limit error, closing it. Thank you.
